File: machine/ts/machine/devices/audio/queue_status_registers.ts
import {
	IO_APU_CMD_CAPACITY,
	IO_APU_CMD_FREE,
	IO_APU_CMD_QUEUED,
} from '../../../spec/bmsx/io';
import type { ApuCommandFifo } from './command_fifo';

export class ApuQueueStatusRegisters {
	public constructor(private readonly commandFifo: ApuCommandFifo) {}

	public static readThunk(context: ApuQueueStatusRegisters, addr: number): number {
		switch (addr) {
			case IO_APU_CMD_QUEUED:
				return context.commandFifo.count;
			case IO_APU_CMD_FREE:
				return context.commandFifo.free;
			case IO_APU_CMD_CAPACITY:
				return context.commandFifo.capacity;
		}
		throw new Error('[APU] Queue-status register read was mapped to an unknown address.');
	}
}

File: machine/cpp/machine/devices/audio/queue_status_registers.cpp
#include "machine/devices/audio/queue_status_registers.h"

#include "spec/bmsx/io.h"
#include "machine/devices/audio/command_fifo.h"

namespace bmsx {

ApuQueueStatusRegisters::ApuQueueStatusRegisters(const ApuCommandFifo& commandFifo)
	: m_commandFifo(commandFifo) {}

u32 ApuQueueStatusRegisters::readThunk(void* context, u32 addr, MappedBusSignals) {
	auto& regs = *static_cast<ApuQueueStatusRegisters*>(context);
	switch (addr) {
		case IO_APU_CMD_QUEUED:
			return regs.m_commandFifo.count();
		case IO_APU_CMD_FREE:
			return regs.m_commandFifo.free();
		case IO_APU_CMD_CAPACITY:
			return regs.m_commandFifo.capacity();
	}
	throw BMSX_RUNTIME_ERROR("[APU] Queue-status register read was mapped to an unknown address.");
}

} // namespace bmsx
